libs/common/RespReadUtils.cs

```diff
@@ -4,156 +4,363 @@
+        /// <returns>True if a length header was successfully read.</returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool ReadLengthHeader(out int len, ref byte* ptr, byte* end, bool isArray = false)
         {
-            number = 0;
-            if (ptr + 3 >= end)
+            len = -1;
+            if (ptr + 3 > end)
                 return false;
 
-            Debug.Assert(*ptr == ':');
+            var readHead = ptr + 1;
+            var negative = *readHead == '-';
 
-            ptr++;
-            number = *ptr++ - '0';
-            while (*ptr != '\r')
+            // String length headers must start with a '$', array headers with '*'
+            if (*ptr != (isArray ? '*' : '$'))
             {
-                Debug.Assert(*ptr >= '0' && *ptr <= '9');
-                number = number * 10 + *ptr++ - '0';
-                if (ptr >= end)
+                RespParsingException.ThrowUnexpectedToken(*ptr);
+            }
+
+            // Special case: '$-1' (NULL value)
+            if (negative)
+            {
+                if (readHead + 4 > end)
+                {
                     return false;
+                }
+                if (*(uint*)readHead == MemoryMarshal.Read<uint>("-1\r\n"u8))
+                {
+                    ptr = readHead + 4;
+                    return true;
+                }
+                readHead++;
             }
-            ptr += 2;
+
+            // Parse length
+            if (!TryReadUlong(ref readHead, end, out var value, out var bytesRead))
+            {
+                return false;
+            }
+
+            if (bytesRead == 0)
+            {
+                RespParsingException.ThrowUnexpectedToken(*ptr);
+            }
+
+            // Validate length
+            len = (int)value;
+
+            if (negative)
+            {
+                RespParsingException.ThrowInvalidStringLength(-len);
+            }
+
+            if (value > int.MaxValue)
+            {
+                RespParsingException.ThrowIntegerOverflow();
+            }
+
+            // Ensure terminator has been received
+            ptr = readHead + 2;
             if (ptr > end)
+            {
                 return false;
+            }
+
+            if (*(ushort*)readHead != MemoryMarshal.Read<ushort>("\r\n"u8))
+            {
+                RespParsingException.ThrowUnexpectedToken(*ptr);
+            }
 
-            Debug.Assert(*(ptr - 1) == '\n');
             return true;
         }
 
-
         /// <summary>
-        /// Read the length of an array of bulk strings
+        /// Read signed 64 bit number
         /// </summary>
-        /// <param name="number"></param>
-        /// <param name="ptr"></param>
-        /// <param name="end"></param>
-        /// <returns></returns>
-        public static bool ReadArrayLength(out int number, ref byte* ptr, byte* end)
+        public static bool Read64Int(out long number, ref byte* ptr, byte* end)
```

Comment:
IIRC, current RESP *at least with Redis* defines all client-to-server comms to be limited to arrays (at the root) and strings (inside the root); see https://redis.io/docs/latest/develop/reference/protocol-spec/#sending-commands-to-a-redis-server

as such, is `:123` etc actually used?

Reply:
Good catch! You are right, client messages should never include integers using this syntax and integer reads are only used internally by some of the bitmap operations. Those reads probably don't need to be explicitly checked for overflows and RESP syntax errors, but given the performance impact seems acceptable I think it's good to keep the checks for internal reads as well.